### ISSUE TITLE
Why does a white background appear when Lapce or Floem Application  is launched, and how can it be eliminated?

### ISSUE THREAD

**Issue by moon989** [lapce/floem#468]
![image](https://github.com/lapce/floem/assets/32634377/c9cbea6c-5049-4f74-8947-50671c440d3c)
**As shown in the above image, there is a white background that appears upon startup; it's very unsightly and affects the visual experience. It is recommended to eliminate it.**

**Comment by moon989** [comment]
@timboudreau   Thank you for your reply； The following is  environment information。

- OS： Win11；
- floem : V0.1.1; 

Running Lapce （All  release version ） also results in a white background appearing, as shown in the image.

**Comment by timboudreau** [comment]
I'm not going to have any ideas about Windows - too many years working at Sun Microsystems where installing Windows was a firing offense...

I spent some time this afternoon attempting and failing to get transparent windows actually working on Mac OS, and ran into similar issues - there's something that really wants to paint the windows white or black (if transparent) that's below the level where Floem does anything (it does pass the transparent flag through to winit - I verified that).

Some googling showed up a set of flags passed to MacOS by winit via wgpu that determine whether the window has an alpha channel at all. Might be an avenue to look at here as well.

Though it may simply be that you need to style your bottommost view with fill_width() and fill_height() and you're just getting the root view (container Floem puts your views in in the window) background because your component doesn't fully cover it.

Try that, in fact.

**Comment by jrmoulton** [comment]
@timboudreau 

specifically for getting transparent windows it has to do with the alpha mode here. Different setting affect different platforms in different ways. 
https://github.com/lapce/floem/blob/1674b0476c4a857322877f765bc8489c0d61d01b/vger/src/lib.rs#L88

**Comment by timboudreau** [comment]
Good to know. I can try to get it happy on Mac OS and Linux/X11 (maybe even Wayland if I want to futz with rebuilding it with different flags - I'm on a circa 2016 Broadwell CPU with a somewhat buggy GPU there). But I have no recourse for testing on Windows.

**Comment by Long0x0** [comment]
I debugged winit on Windows and found the same problem, just after [the window is created](https://github.com/lapce/winit/blob/08a3dfc1ca8cac27c4a404e2eed92dfb77167821/src/platform_impl/windows/window.rs#L1353). 

According to MS docs, we need to set window's background in [`register_window_class`](https://github.com/lapce/winit/blob/08a3dfc1ca8cac27c4a404e2eed92dfb77167821/src/platform_impl/windows/window.rs#L1394) or process the `WM_ERASEBKGND` message.

https://learn.microsoft.com/en-us/windows/win32/api/winuser/ns-winuser-wndclassa

**Comment by moon989** [comment]
How can I solve this problem? Is there anyone who can provide suitable suggestions?

**Comment by jrmoulton** [comment]
@moon989 the default background color (the clear color) is set by the renderer

https://github.com/lapce/floem/blob/02205d851df89e82ef297e3ceb305551f37ee944/vger/src/lib.rs#L573

There is probably a way to add a color to the window config and get it passed down to this point

**Comment by panekj** [comment]
> There is probably a way to add a color to the window config and get it passed down to this point

It's already transparent, there is no need to pass anything

**Comment by jrmoulton** [comment]
> It's already transparent, there is no need to pass anything

But that's the issue. We don't always want it to be transparent. 

Setting this clear color to a specific value will make the window clear to that color when initializing (I think).

**Comment by panekj** [comment]
> We don't always want it to be transparent.

We do, otherwise you will get either different colours between the app itself and what you provide in `View`.
I specifically made that change so it doesn't require any colour configuration.

**Comment by panekj** [comment]
I'm pretty sure the issues with white window flash at the start are related to the backend used and/or transparency mode used.
In particular transparency behaves weirdly between modes and `Auto` is not handling all cases.
https://github.com/lapce/floem/commit/3b1cb03e49c40f53b5f2bae4d28cc0fb1881aa20

**Comment by moon989** [comment]
This bug is also exist in  Lapce  all versions   Release .

**Comment by panekj** [comment]
Lapce uses Floem, so all issues in Floem will be present in Lapce

**Comment by timboudreau** [comment]
FWIW, on Mac OS with a transparent window background set and the flag for a transparent window set when creating the window, you get a window with a black background.

So I think something is amiss with window initialization, though it may or may not be related to this issue on Windows. I've tried patching it a few different ways but have not found a fix yet.

In researching it, I ran across a comment on a bug report for something using winit saying that winit's implementation of NSView was missing some properties introduced more recently or trying to use properties that are no longer used. I don't know the accuracy of that but it might be a thing you check for solving window background issues Mac OS.

**Comment by moon989** [comment]
This bug also exist in  the latest version of lapce . I'm trying to build an application on Windows using floem, but this problem, if not resolved, is very user-unfriendly to  user. 
![image](https://github.com/user-attachments/assets/f5822a7d-2427-4904-a549-6e52bfffb7e7)

**Comment by moon989** [comment]
![image](https://github.com/user-attachments/assets/4aeaf1b3-d8a4-4574-9829-e08fb9659edc)
It's not a bug ?

**Comment by dzhou121** [comment]
> ![image](https://private-user-images.githubusercontent.com/32634377/360225780-4aeaf1b3-d8a4-4574-9829-e08fb9659edc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQzMTk3MzQsIm5iZiI6MTcyNDMxOTQzNCwicGF0aCI6Ii8zMjYzNDM3Ny8zNjAyMjU3ODAtNGFlYWYxYjMtZDhhNC00NTc0LTk4MjktZTA4ZmI5NjU5ZWRjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODIyVDA5MzcxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA3Y2YyZDQyYmU3YWZjZWMzYzllMDVlYmE0Y2JkNDRkMTY2MTQ0MzU2OWQyNzRiODc5OWMzMGQ1ODUyZmU3MzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.6BeLtJ25IUlAhf3p16Z6kEJdqazu7PFTUpX8TO4dxTg) It's not a bug ?

It is a bug, but we haven't figured out how to fix it yet.

**Comment by Long0x0** [comment]
Finally fixed this on winit: https://github.com/Long0x0/winit/commit/23722eaeb72778245288e6a4d696902be755ff33. The window will remain transparent until drawn by floem. Could you check if this works?

**Comment by moon989** [comment]
> Finally fixed this on winit: [Long0x0/winit@23722ea](https://github.com/Long0x0/winit/commit/23722eaeb72778245288e6a4d696902be755ff33). The window will remain transparent until drawn by floem. Could you check if this works?

Great, this has fixed the issue, and I've tested it on Windows 11. Hope it gets merged into the main branch soon.

**Comment by moon989** [comment]
Is there a plan to fix this bug?

**Comment by dzhou121** [comment]
> Is there a plan to fix this bug?

It's currently blocked by this PR

**Comment by Long0x0** [comment]
After investigating many issues in rust-windowing/winit, I created an alternative PR to solve this.
